---
---

The philosophy of 'BCH709 Introduction to Bioinformatics' is to teach researchers basic concepts, skills, and tools for working with data so that student can get their own bioinformatics work done without pain. This teaches concept of bioinformatics analysis and Next Generation Sequencing for genomics including: hands-on bioinformatics, understanding genomics through bioinformatics, use of command-line in multiple platform, use of command-line tools to analyze sequencing data, and connecting to and using cloud computing.


> ## Course description
> An introduction to the science of genomics and bioinformatics. Includes genome sequencing; database techniques; structural, comparative, and evolutionary genomics; and gene expression analysis.
{: .prereq}

> ## Class Room location
> ![Classroom location](./fig/classroom_location.png){: width="50%" height="50%"}
> 
> PMB 104D (inside of Savitt Medical Library)  
> [Google Map](https://goo.gl/maps/GGx8NTwfyi8GZFpb8)
{: .prereq}

> ## Class Schedule
> TuTh 9:00AM - 10:15AM  
> Aug 26, 2019 - Dec 10, 2019
{: .prereq}

> ## Prerequisites
> Online introduction to Linux. Students must complete one of the following online tutorials (or both) before class begins.
> - [Code academy's Intro to Unix](https://www.codecademy.com/learn/learn-the-command-line "Code academy")
> - Computer with ethernet port or wifi (If in case you bring your **desktop**, please do not bring your monitor. we have a monitor in our classroom)
>
> - Terminal software
> - UNR affilated email **\<ID\>@unr.edu or \<ID\>@nevada.unr.edu**
{: .callout}





> ## Frequently Asked Questions
> Read our [FAQ](/genomics-workshop/faq/) to learn more about Data Carpentry's Genomics workshop, as an Instructor or a workshop host.
{: .prereq}

> ## Getting Started
>
> This lesson assumes that learners have no prior experience with the tools covered in the workshop. 
> However, learners are expected to have some familiarity with biological concepts,
> including the 
> concept of genomic variation within a population. Participants should bring their own laptops and plan to participate actively. 
> 
> To get started, follow the directions in the [Setup](setup.html) tab to 
> get access to the required software and data for this workshop.
> 
{: .prereq}

> ## Data
> 
> This workshop uses data from a long term evolution experiment published in 2016: [Tempo and mode of genome evolution in a 50,000-generation experiment](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4988878/) by Tenaillon O, Barrick JE, Ribeck N, Deatherage DE, Blanchard JL, Dasgupta A, Wu GC, Wielgoss S, Cruveiller S, MÃ©digue C, Schneider D, and Lenski RE. (doi: 10.1038/nature18959)
>
> All of the data used in this workshop can be [downloaded from Figshare](https://figshare.com/articles/Data_Carpentry_Genomics_beta_2_0/7726454). 
> More information about this data is available on the [Data page](https://datacarpentry.org/organization-genomics/data/).
{: .prereq} 

# Class Overview 

| Week    | Tuesday | Thursday|
| ------- | ---------- |---------- |
|Week1|Intro Bioinformatics|Linux Enviroment and command line|
|Week2|Linux Enviroment and bioawk|Linux Enviroment and Cloud|
|Week3|Github|Compile|
|Week4|Sequencing methods and strategies|RNA-Seq|
|Week5|Transcriptome assembly|Sequence similarity searching|
|Week6|Functional annotation|Database|
|Week7|Introduction of R R plotting (Heatmap & others)|DEG analaysis
|Week8|DESeq2 / EdgeR|Tong Zhou (his research topic)|
|Week9|Quiz: Transcriptome & Database|Gene family analysis and phylogenetics|
|Week10|Genome assembly|Genome assembly|
|Week11|Genome assembly|How to annotate genomes|
|Week12|Genome assembly and annotation|Genome structure|
|Week13|DEG analaysis|Transcriptome analaysis
|Week14|Variant analaysis|Thanks giving|
|Week15|Enrichment analysis|Presentation|
|Week16|Final exam|

# Optional Additional Meeting

| Lesson | Overview |
| ------- | -------- |
| [Intro to R and RStudio for Genomics](https://datacarpentry.org/genomics-r-intro/) | Use R to analyze and visualize between-sample variation. |

# Teaching Platform
This workshop is designed to be run on pre-imaged Amazon Web Services (AWS)
instances. All the software and data used in the workshop are hosted on an Amazon Machine Image (AMI).
If you want to run your own instance of the server used for this workshop, follow the directions in the [Setup](setup.html) tab.
