---
---

The philosophy of 'BCH709 Introduction to Bioinformatics' is to teach researchers basic concepts, skills, and tools for working with data so that student can get their own bioinformatics work done without pain. This teaches concept of bioinformatics analysis and Next Generation Sequencing for genomics including: hands-on bioinformatics, understanding genomics through bioinformatics, use of command-line in multiple platform, use of command-line tools to analyze sequencing data, and connecting to and using cloud computing.


> ## Course description
> An introduction to the science of genomics and bioinformatics. Includes genome sequencing; database techniques; structural, comparative, and evolutionary genomics; and gene expression analysis.
{: .prereq}

> ## Class Room location
> ![Classroom location](./fig/classroom_location.png){: width="50%" height="50%"}
> 
> PMB 104D (inside of Savitt Medical Library)  
> [Google Map](https://goo.gl/maps/GGx8NTwfyi8GZFpb8)
{: .prereq}

> ## Class Schedule
> TuTh 9:00AM - 10:15AM  
> Aug 26, 2019 - Dec 10, 2019
{: .prereq}

> ## Prerequisites
> Online introduction to Linux. Students must complete one of the following online tutorials (or both) before class begins.
> - [Code academy's Intro to Unix](https://www.codecademy.com/learn/learn-the-command-line "Code academy")
> - Computer with ethernet port or wifi (If in case you bring your **desktop**, please do not bring your monitor. we have a monitor in our classroom)
>
> - Terminal software
> - UNR affilated email **\<ID\>@unr.edu or \<ID\>@nevada.unr.edu**
{: .callout}

> ## Getting Started
>
> This lesson assumes that learners have no prior experience with the tools covered in the workshop. 
> However, learners are expected to have some familiarity with biological concepts,
> including the 
> concept of genomic variation within a population. Participants should bring their own laptops and plan to participate actively. 
> 
> To get started, follow the directions in the [Setup](setup.html) tab to 
> get access to the required software and data for this workshop.
> 
{: .prereq}
<!-- 
> ## Data
> 
> This workshop uses data from a long term evolution experiment published in 2016: [Tempo and mode of genome evolution in a 50,000-generation experiment](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4988878/) by Tenaillon O, Barrick JE, Ribeck N, Deatherage DE, Blanchard JL, Dasgupta A, Wu GC, Wielgoss S, Cruveiller S, Médigue C, Schneider D, and Lenski RE. (doi: 10.1038/nature18959)
>
> All of the data used in this workshop can be [downloaded from Figshare](https://figshare.com/articles/Data_Carpentry_Genomics_beta_2_0/7726454). 
> More information about this data is available on the [Data page](https://datacarpentry.org/organization-genomics/data/).
{: .prereq} 
-->

## Class Overview 

| Week    | Tuesday | Thursday|
| ------- | ---------- |---------- |
|Week1|Intro Bioinformatics|Linux Enviroment and command line|
|Week2|Linux Enviroment and bioawk|Linux Enviroment and Cloud|
|Week3|Github|Compile|
|Week4|Sequencing methods and strategies|RNA-Seq|
|Week5|Transcriptome assembly|Sequence similarity searching|
|Week6|Functional annotation|Database|
|Week7|Introduction of R &  R plotting ([Tong Zhou PhD](http://tongzhoulab.org/))|DEG analaysis
|Week8|DESeq2 / EdgeR|Special topics (Tong Zhou PhD)|
|Week9|Quiz: Transcriptome & Database|Gene family analysis and phylogenetics ([David Alvarez-Ponce, PhD](https://genomeevol.wordpress.com/))|
|Week10|Genome assembly|Genome assembly|
|Week11|Genome assembly|How to annotate genomes|
|Week12|Genome assembly and annotation|Genome structure|
|Week13|DEG analaysis|Transcriptome analaysis
|Week14|Variant analaysis|[Thanksgiving](https://giphy.com/search/dance)|
|Week15|Enrichment analysis|Presentation|
|Week16|Final exam|

## Optional Additional Meeting

### _Research Computing Hackathon_  ([Hosted by HPC team](https://www.unr.edu/research-computing/hpc))  
[Mathewson-IGT Knowledge Center, MIKC 414](https://events.unr.edu/mathewson-igt_knowledge_center_508#.XVyb3OhKiiM)  
Every Friday at 2:00pm to 4:00pm

[Hackathons](https://en.wikipedia.org/wiki/Hackathon) provide a space for hands-on training and solution development within a Research Computing environment at the University. This is also a place to get clarification on questions/concerns regarding the HPC environment. Please bring problems to challenge the HPC team, the Office of Information Technology, and research colleagues. If you don't need help, we still encourage you to attend and share your time and expertise with those in need of assistance. You don’t need to be an expert to attend a hackathon. Individuals at all computing skill levels are welcome!

### _Meeting with Won Yim_  
[Howard Molecular Science 216](https://goo.gl/maps/o41BMmcawsTPoES57)  
After Thursday class, you can schedule a meeting before noon. We are not providing any financial aid, but we can disccuss your research bottleneck in terms of bioinformatics perspective. **Please schedule after Tuesday class.**


> ## Frequently Asked Questions
> Read our [FAQ](./_episodes/FAQ/FAQ.md). Currently this page is empty but we will build it through the class.
{: .prereq}

>## Teaching Platform
This workshop is designed to be run on [Unix-base system](https://en.wikipedia.org/wiki/Unix) such as 
Ubuntu, mac etc. All the software and data used in the class will be open source All example data will be hosted on an Google Cloud Service. If you want to know how to use Unix-base system on your computer, please follow the directions in the [Setup](setup.html) tab.
{: .prereq}


~~~
The website theme was adapted from the original by [Data Carpentry](https://datacarpentry.org/). The infrastructure, including adventure-time and docker-browser-server, was built by @maxogden and @mafintosh. The setup of this app was based on the get-dat adventure. This adventure was made by Richard Smith-Unna. The lecture materials were crafted by Won Yim. This work is licensed under a Creative Commons 4.0 International License. Creative Commons License
~~~

